src/chunk_append/exec.c

```diff
@@ -355,17 +385,48 @@ choose_next_subplan(ChunkAppendState *state)
 		 * bms_next_member will return -2 (NO_MATCHING_SUBPLANS) if there are
 		 * no more members
 		 */
-		state->current = bms_next_member(state->valid_subplans, state->current);
+		return bms_next_member(state->valid_subplans, last_plan);
 	}
 	else
 	{
-		state->current++;
+		int next_plan = last_plan + 1;
+
+		if (next_plan >= state->num_subplans)
+			return NO_MATCHING_SUBPLANS;
 
-		if (state->current >= state->num_subplans)
-			state->current = NO_MATCHING_SUBPLANS;
+		return next_plan;
 	}
 }
 
+static void
+choose_next_subplan_local(ChunkAppendState *state)
+{
+	state->current = get_next_subplan(state, state->current);
+}
+
+static void
+choose_next_subplan_for_leader(ChunkAppendState *state)
+{
+	state->current = NO_MATCHING_SUBPLANS;
```

Comment:
I think this is OK if we cannot end up in a state where we have parallel execution with a leader but no workers. Is that possible? I am a bit worried about being unable to make progress in this case.

Reply:
since postgres allows this i dont think this can happen, but i can make leader participation a GUC similar to postgres